1.21.11 Update GameRule API

ekulxam opened the pull request:
Confirmed that Game Test, Client Game Test, and Auto Test Server all succeeded.

**Changes**
- Port to 1.21.11
- Remove every custom rule class. Everything is now handled by Minecraft's `GameRule`.
- Use local mappings (I will revert this commit once https://github.com/FabricMC/yarn/pull/4367 is merged)
- Fix test(mod) to the best of my ability
- TAW all register methods in `GameRules`
- Change callback seems to no longer exist, so all references to that no longer exist
- GameRuleFactory now creates `GameRule`s and not the `Type`
- GameRuleFactory defaults to `MISC` for the rule `Category`

Do we still need `CustomGameRuleCategory`? `Category` is no longer an enum and can be properly instantiated using `method_76160`, which somewhat defeats the purpose of having `CustomGameRuleCategory` in the first place. I only kept it because `CustomGameRuleCategory`'s name / translation key can be configured, but I'm willing to hear other viewpoints.

I don't know what is happening in `GameRuleManagementHandlerImplMixin`, but I think commenting out the updateRule handler was correct, as it doesn't seem to be different from what vanilla is doing now.

As a side note, my import order is super broken, and I don't know how to fix it.

ekulxam committed:
Update yarn with local version

ekulxam committed:
Enable gamerule api

ekulxam committed:
Fix GameRules

The mappings did fairly well
Behold my FAPI mixin hacks

ekulxam committed:
Fix yarn

ekulxam committed:
spotlessApply

ekulxam committed:
Revert CustomGameRuleCategory to final class

ekulxam committed:
Fix checkstyle

offbeat-stuff commented:
In 25w44a gamerules update callbacks happen in a function in server world.

Can you add an event to easily register a callback for your gamerules?

ekulxam committed:
Add change callback

Earthcomputer commented on fabric-game-rule-api-v1/src/test/java/net/fabricmc/fabric/test/gamerule/GameRuleManagementHandlerImplTest.java:
```diff
@@ -120,19 +133,19 @@ void testUpdateVanillaInt() {
 		GameRuleManagementHandler handler = new GameRuleManagementHandlerTestImpl(server, MANAGEMENT_LOGGER);
 
-		GameRuleRpcDispatcher.TypedRule result = handler.updateRule(new GameRuleRpcDispatcher.UntypedRule("randomTickSpeed", "123"), CONNECTION_ID);
+		GameRuleRpcDispatcher.class_12254<Integer> result = handler.updateRule(new GameRuleRpcDispatcher.class_12254<>(GameRules.RANDOM_TICK_SPEED, 123), CONNECTION_ID);
```
Might be worth it to update yarn so these are named

ekulxam replied:
Even after working with this for a few hours, I'm still not sure what I would name this if given the chance. TypedRule seems outdated becaues this class has basically replaced both TypedRule and UntypedRule, so perhaps a better name would be `RuleValueContainer` or `RuleValuePair`?

apple502j replied:
Key-value pair is an entry, e.g. RegistryEntry.

(Hold(er) on, that's not how it's going to be named anymore...)

Earthcomputer replied:
I just assumed that it was mapped in yarn already tbh

ekulxam committed:
class_12254Mixin

ekulxam committed:
Fix event test

ekulxam committed:
Reorder change callback, fix mixin names

Boo validateMixinNames doesn't let me do the inner class thing

ekulxam committed:
Extremely cursed server creation

ekulxam commented:
e561067 can be reverted as necessary because it will probably be a pain to maintain. I spent the better part of a few hours figuring out the minimum `mock`s required to create a MinecraftDedicatedServer, so I understand if it is removed.

Earthcomputer commented:
@ekulxam why not make it into a game test?

ekulxam commented:
Amazing, can't believe I missed that. I suppose I'm learning game tests now.

ekulxam committed:
Revert "Extremely cursed server creation"

This reverts commit e56106723c2b4513f633d3bb1fac64ebc9794116.

ekulxam committed:
Replaced cursed mocking with gametest

modmuss50 commented on build.gradle:
```diff
@@ -119,6 +119,10 @@ allprojects {
 		enabled = false
 	}
 
+	repositories {
```
remove

modmuss50 commented on fabric-game-rule-api-v1/src/main/java/net/fabricmc/fabric/api/gamerule/v1/GameRuleEvents.java:
```diff
@@ -0,0 +1,55 @@
+public final class GameRuleEvents {
+	private GameRuleEvents() { }
+
+	public static final Event<ValueUpdate> CHANGED_CALLBACK = EventFactory.createArrayBacked(
```
Would this be better as something along the following lines:

`<T> Event<ValueUpdate<T>> changedCallback<GameRule<T> gameRule)`

basically a static method that returns you the event for a specific gamerule, this way it can be typesafe.

modmuss50 commented on fabric-game-rule-api-v1/src/main/java/net/fabricmc/fabric/api/gamerule/v1/GameRuleRegistry.java:
```diff
@@ -16,52 +16,51 @@
+ * contains many registration methods not found here.
  */
 public final class GameRuleRegistry {
+	public static final Logger LOGGER = LoggerFactory.getLogger(GameRuleRegistry.class);
```
private

modmuss50 replied:
Doesnst seem used so remove.

modmuss50 reviewed (commented):
Looking good, the main thing is the value changed event, I think it can be made typesafe quite easily.

apple502j commented on fabric-game-rule-api-v1/src/main/java/net/fabricmc/fabric/api/gamerule/v1/GameRuleEvents.java:
```diff
@@ -0,0 +1,55 @@
+		 * @param server the server
+		 * @see MinecraftServer#onGameRuleUpdated(GameRule, Object)
+		 */
+		void accept(
```
FAPI event handlers generally start with `on`

apple502j commented on fabric-game-rule-api-v1/src/main/java/net/fabricmc/fabric/api/gamerule/v1/GameRuleFactory.java:
```diff
@@ -18,301 +18,168 @@
  *
- * <p>To register a game rule in a custom category, {@link GameRuleRegistry#register(String, CustomGameRuleCategory, GameRules.Type)} should be used.
+ * <p>To register a game rule in a custom category, {@link GameRuleRegistry#register(String, GameRule, CustomGameRuleCategory)} should be used.
+ * Alternatively, cast the {@link GameRule} to {@link RuleCategoryExtensions} and call {@link RuleCategoryExtensions#fabric_setCustomCategory(CustomGameRuleCategory)}.
```
Remove, never link to impl in javadoc.

apple502j commented on fabric-game-rule-api-v1/src/main/java/net/fabricmc/fabric/api/gamerule/v1/GameRuleFactory.java:
```diff
@@ -18,301 +18,168 @@
+ * For example, to register a game rule that is an integer where the default value is 1 and the acceptable values are between 0 and 10, one would use the following:
  * <blockquote><pre>
- * public static final GameRules.Key&lt;GameRules.IntRule&gt; EXAMPLE_INT_RULE = GameRuleRegistry.register("exampleIntRule", GameRules.Category.UPDATES, GameRuleFactory.createIntRule(1, 10));
+ * public static final GameRule&lt;Integer&gt; EXAMPLE_INT_RULE = GameRules.registerIntRule("example_int_rule", Category.MISC, 1, 0, 10);
```
Remove, as the register methods in GameRules take String, not Identifier.

apple502j commented on fabric-game-rule-api-v1/src/main/java/net/fabricmc/fabric/api/gamerule/v1/GameRuleFactory.java:
```diff
@@ -18,301 +18,168 @@
  *
  * @see GameRuleRegistry
  */
 public final class GameRuleFactory {
```
If we are completely revamping the class, I think a redesign might be better. Why not a builder, for example?

Also rename the class since this is not a real factory (factory in Java is a functional interface for creating an instance)

apple502j reviewed (changes requested):
Stopping mid-review, this clearly needs more work.

ekulxam committed:
GameRuleBuilder, typesafe event

ekulxam commented:
For `MinecraftServerMixin#invokeChangeCallbacks`, do we want to be able to not invoke anything if the event doesn't exist? In the current state, a new Event will be created for every GameRule after that rule has been changed. Is it worth it to only invoke Events for already existing ones (i.e. don't call computeIfAbsent for the invoker).

modmuss50 commented on fabric-game-rule-api-v1/src/main/java/net/fabricmc/fabric/api/gamerule/v1/GameRuleEvents.java:
```diff
@@ -0,0 +1,67 @@
+	private GameRuleEvents() {
+	}
+
+	private static final Map<GameRule<?>, Event<ValueUpdate<?>>> VALUE_UPDATES = new HashMap<>();
```
Move this to an impl class.

modmuss50 replied:
Can this be an IdentityHashMap?

modmuss50 commented on fabric-game-rule-api-v1/src/main/java/net/fabricmc/fabric/api/gamerule/v1/GameRuleEvents.java:
```diff
@@ -0,0 +1,67 @@
+					changedCallback.onGameRuleUpdated(value, server);
+				}
+			});
+		});
```
Move this to an impl class, you should only create the event when somone uses the api. When you invoke the event you can get null back from the map and not invoke anything.

modmuss50 commented on fabric-game-rule-api-v1/src/main/java/net/fabricmc/fabric/mixin/gamerule/MinecraftServerMixin.java:
```diff
@@ -16,18 +16,20 @@
+	@WrapMethod(method = "onGameRuleUpdated")
+	private <T> void invokeChangeCallbacks(GameRule<T> rule, T value, Operation<Void> original) {
+		original.call(rule, value);
+		GameRuleEvents.changeCallback(rule).invoker().onGameRuleUpdated(value, (MinecraftServer) (Object) this);
```
I dont think this is the correct place to inject this, as you will miss all of the gamerule changes that happened without a server. (See TestWorldBuilderImpl)

ekulxam committed:
GameRuleEventsImpl

ekulxam committed:
Merge branch '1.21.11' into 1.21.11

apple502j commented:
Still some issues, @ekulxam can I just make a PR to your branch?

apple502j committed:
API improvements

apple502j committed:
Further API changes

- Removed GameRuleRegistry since we have a builder providing the necessary features
- Move game rule update event back to MinecraftServer
- Rename builder methods

apple502j committed:
Improve validation, rename misleading method

apple502j committed:
Ensure enum default is supported

apple502j commented:
@ekulxam PRed, https://github.com/ekulxam/fapi/pull/1

ekulxam committed:
Bump yarn version to map more of rules

ekulxam committed:
Fix yarn conflicts

ekulxam committed:
Fix checkstyle and other issues from yarn update

ekulxam committed:
Fix rule widgets
